### PR TITLE
 D2IQ-71316 - dependency picker in jobs form

### PR DESCRIPTION
**Pull request by pierrebeitz**
when adding a job, you can now select dependency-jobs in the general tab when
your cluster version is >= 2.2.0-alpha.

screencast 2020-08-26 22-30-40

### CODE REVIEW

File: tests/pages/jobs/JobCreateForm-cy.js [hunk 1]
```diff
@@ -55,13 +55,17 @@ describe("Job Create Form", () => {
...
-        cy.root()
-          .getFormGroupInputFor("Job ID *")
-          .type(`{selectall}${"simple"}`);
+        cy.getFormGroupInputFor("Job ID *").retype("simple");
 
         // Error badge disappears
         getActiveTabErrorBadge().should("not.be.visible");
       });
+
+      it("adds and removes a dependency", () => {
```
Comment (janisz) [inline tests/pages/jobs/JobCreateForm-cy.js:64]: How about adding more than one dependecy?

**Review by janisz — COMMENTED** [review]
Looks good but I have some questions.
:+1: For two commits (maybe to PRs will be better).

File: locale/zh/messages.json [hunk 2]
```diff
@@ -1131,6 +1132,7 @@
   "This host declined an offer because it has scarce resources (e.g. GPUs).": "",
   "This host port will be accessible as an environment variable called '$PORT{index}'. <0>More information</0>.": "此主机端口可以作为一个名为 '$PORT{index}' 的环境变量进行访问。<0>更多信息</0>.",
   "This host port will be accessible as an environment variable called {environmentVariableName}. <0>More information</0>.": "此主机端口可以作为一个名为 {environmentVariableName} 的环境变量进行访问。<0>更多信息</0>.",
+  "This job will run whenever its last successful run is older than all of its dependencies' latest successful runs.": "",
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply locale/zh/messages.json:1135]: chinese translations will always be empty at first. once in a while we give that file to an agency that translates all the new entries.

File: plugins/jobs/src/js/components/form/GeneralFormSection.tsx [hunk 3]
```diff
@@ -248,27 +184,25 @@ export default class GeneralFormSection extends React.Component<GeneralProps> {
...
-                    wrapText={true}
-                  >
-                    <InfoTooltipIcon />
-                  </Tooltip>
-                </FormGroupHeadingContent>
-              </FormGroupHeading>
-            </FieldLabel>
             <FieldAutofocus>
-              <FieldTextarea
+              <Textarea
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply plugins/jobs/src/js/components/form/GeneralFormSection.tsx:188]: same reason why we changed all the `FieldInput` to `TextInput`: we want to eventually have all components on ui-kit.

currently the ui has lots of "legacy" components that have been moved to ui-kit in the past. they look exactly the same and sometimes even have the same interface. i try to be a good citizen here and convert stuff as i touch it. 
whenever a component is reasonably close to being extincted i'll go all in and change all of the remaining occurrences. sometimes that's not easy because of past design decisions.

so TL;DR:
the boy scout approach :)

File: src/styles/components/modal/styles.less [hunk 4]
```diff
@@ -1,5 +1,10 @@
 & when (@modal-enabled) {
 
+  // diplay ui-kit popups and tooltips in front of .modal-wrapper
+  [data-popper-placement] {
+    z-index: 1010 !important;
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply src/styles/components/modal/styles.less:5]: it is. we want ui-kits tooltips (which seem to have a `z-index` of `2`) in front of our modal-wrapper, which has a `z-index` of `1000`. if i just turn that down, there's stuff from all over the place bleeding into the modal on different pages. i don't want to go down that rabbit hole just now, but still enable the usage of ui-kit tooltips on our modals (and we "abuse" modals to hold the service/jobs form unfortunately).

File: tests/pages/jobs/JobCreateForm-cy.js (hunk 1 above)
Comment (pierrebeitz) [reply tests/pages/jobs/JobCreateForm-cy.js:64]: yeah, fair point!

**Comment by danielone** [comment]
@pierrebeitz  what's the text in the tooltip
otherwise looks great

File: plugins/jobs/src/js/components/JobsForm.tsx [hunk 5]
```diff
@@ -42,8 +42,9 @@ import JobErrorTabPathRegexes from "../constants/JobErrorTabPathRegexes";
 
 const ServiceErrorPathMapping: any[] = [];
 
-const JSONEditor = React.lazy(() =>
-  import(/* webpackChunkName: "jsoneditor" */ "#SRC/js/components/JSONEditor")
+const JSONEditor = React.lazy(
+  () =>
+    import(/* webpackChunkName: "jsoneditor" */ "#SRC/js/components/JSONEditor")
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply plugins/jobs/src/js/components/JobsForm.tsx:47]: i thought that prettyCron was a so small dependency that it's not worth lazy loading that component - so i simplified stuff a bit. i can definitely change that back.

**Comment by pierrebeitz** [comment]
@danielone those are the texts: 

* `No more jobs to choose from.`
* `This job will run whenever its last successful run is older than all of its dependencies' latest successful runs.`

I've not run them by Vinod as i did not know there was such a process. Changing those is pretty trivial.

**Comment by vinodkone** [comment]
The text looks good to me.

**Comment by danielone** [comment]
Thanks @pierrebeitz 

Is this in the tooltip? "This job will run whenever its last successful run is older than all of its dependencies' latest successful runs."

**Comment by danielone** [comment]
@pierrebeitz 

"No more jobs to choose from." => "No more jobs to select."

@mperrotti do we have a standard for TextInputWithBadges when no additional options to select?

**Comment by danielone** [comment]
@pierrebeitz 


current text"
"This job will run whenever its last successful run is older than all of its dependencies' latest successful runs."

Leading with "This job" is a bit ambiguous if taking about new job or 1 or more jobs defined that can be defined here (also using this implies that only 1 dependency can be added."

proposed text:
"1 or more jobs that need to first successfully run before this new job can be run."

using more present/active tense and tone.

please comment on..

**Comment by pierrebeitz** [comment]
@danielone i'll take that into consideration. as we're in new/edit mode here, we should not use your proposal as-is. also it does not quite match the behaviour we're expecting.

here's the [design doc](https://docs.google.com/document/d/1E06TBKMREyfgVAERfcxZfHfXeqWLIwu5Pcihg_Ju6Dk/edit#) in case you want to have a look at the exact behaviour. here's a TL;DR: `Dependent jobs should run if their last successful run is older than all of their parent jobs.`

**Comment by danielone** [comment]
> @danielone i'll take that into consideration. as we're in new/edit mode here, we should not use your proposal as-is. also it does not quite match the behaviour we're expecting.
> 
> here's the [design doc](https://docs.google.com/document/d/1E06TBKMREyfgVAERfcxZfHfXeqWLIwu5Pcihg_Ju6Dk/edit#) in case you want to have a look at the exact behaviour. here's a TL;DR: `Dependent jobs should run if their last successful run is older than all of their parent jobs.`
> 
> i think the discussion should also be about whether we call things "dependencies and dependants" or "parent and children" or even something else. not sure which has the least potential for confusion.

@pierrebeitz  I had a hunch I didn't totally get it right thus why to iterate with you on =) 
1-  Agree on the verbiage aspect. what is currently being used in DC/OS? Is there any such nomenclature used in DC/OS?
2- Then, text needs a pass as this is being done as part of new/edit mode for a job, the goal of the tooltip also needs to convey the *impact to the new job* being created here by specifying dependencies firstly so the user understands the impact to the parent and the children. 

I'll take a generic pass that can be filled in for iteration on.

**Comment by pierrebeitz** [comment]
> 1- Agree on the verbiage aspect. what is currently being used in DC/OS? Is there any such nomenclature used in DC/OS?

unfortunately we don't have this concept yet.

> 2- Then, text needs a pass as this is being done as part of new/edit mode for a job, the goal of the tooltip also needs to convey the impact to the new job being created here by specifying dependencies firstly so the user understands the impact to the parent and the children.

we're in a create/edit-modal for "the job". we now can specify dependencies of this job. the impact to the new job is that it runs "whenever its last successful run is older than all of its dependencies' latest successful runs". for the dependants that means that their runs might trigger a run of this job. i'm not sure what more or how better to convey. 🤔

**Comment by danielone** [comment]
1 - i would recommend to use AWS verbiage
job dependices
dependent job 
https://docs.aws.amazon.com/batch/latest/userguide/job_dependencies.html

**Comment by danielone** [comment]
2- thanks for the copy from scope doc but again it is a bit hard to follow. Here is pass based on AWS which if same functionality then has same net effect:

"Specify dependencies on other jobs that must succeed before this job can run."

simple and clear IMO.

**Comment by janisz** [comment]
I tried to test it but It looks like metronome is not ready yet

Run on 

```
$ curl -k 'https://54.89.204.197/service/metronome/v1/jobs' -X POST --data-raw '{"id":"test","run":{ "dependencies": [{ "id": "dep" }], "cpus":1,"mem":128,"disk":0,"gpus":0,"cmd":"sleep 100","maxLaunchDelay":3600,"restart":{"policy":"NEVER"}}}'

{"message":"Object is not valid","details":[{"path":"/run","errors":["Additional properties are not allowed, but found properties 'dependencies'."]}]}% 
```

**Comment by janisz** [comment]
I found one issue. Autocomplete does not filter current job and suggest it as a dependency when editing existing job.


screen-recorder-wed-sep-16-2020-21-30-05

**Comment by DominikDary** [comment]
> I found one issue. Autocomplete does not filter current job and suggest it as a dependency when editing existing job.
> 
> screen-recorder-wed-sep-16-2020-21-30-05

Thanks for finding it. Can you please create a bug in jira and then we can triage it there with other integration issues we will find during the exploratory testing?

**Comment by janisz** [comment]
https://jira.d2iq.com/browse/DCOS_OSS-5990
